File: FSense/Features/FlowerCard/Models/FlowerProduct.swift
import Foundation

/// Product card model for "Find Flowers" search results (maps to ShopCard on backend)
struct FlowerProduct: Identifiable, Codable, Equatable {
    let productId: String
    let name: String
    let price: String
    let priceValue: Double?
    let currency: String?
    let imageUrl: String?
    let vendor: String
    let buyUrl: String
    let city: String?

    var id: String { productId }

    enum CodingKeys: String, CodingKey {
        case productId = "product_id"
        case name
        case price
        case priceValue = "price_value"
        case currency
        case imageUrl = "image_url"
        case vendor
        case buyUrl = "buy_url"
        case city
    }

    var imageURL: URL? {
        guard let imageUrl else { return nil }
        return URL(string: imageUrl)
    }

    var purchaseURL: URL? {
        URL(string: buyUrl)
    }

    /// Currency symbol based on currency code
    var currencySymbol: String {
        switch currency?.uppercased() {
        case "RUB": return "₽"
        case "CAD": return "CA$"
        case "USD", .none: return "$"
        default: return currency ?? "$"
        }
    }
}

/// Response wrapper for product search
struct FlowerSearchResponse: Codable {
    let success: Bool
    let query: String
    let provider: String?
    let products: [FlowerProduct]
    let error: String?
    let cachedAt: String?  // ISO timestamp when results were cached

    enum CodingKeys: String, CodingKey {
        case success
        case query
        case provider
        case products
        case error
        case cachedAt = "cached_at"
    }
}

// MARK: - Mock Data

extension FlowerProduct {
    static let mock = FlowerProduct(
        productId: "mock_001",
        name: "Dozen Red Roses Bouquet",
        price: "$49.99",
        priceValue: 49.99,
        currency: "USD",
        imageUrl: "https://images.unsplash.com/photo-1518882605630-8eb572299d16?w=400",
        vendor: "1-800-Flowers",
        buyUrl: "https://www.1800flowers.com",
        city: "New York"
    )

    static let mockProducts: [FlowerProduct] = [
        FlowerProduct(
            productId: "mock_001",
            name: "Dozen Red Roses Bouquet",
            price: "$49.99",
            priceValue: 49.99,
            currency: "USD",
            imageUrl: "https://images.unsplash.com/photo-1518882605630-8eb572299d16?w=400",
            vendor: "1-800-Flowers",
            buyUrl: "https://www.1800flowers.com",
            city: "New York"
        ),
        FlowerProduct(
            productId: "mock_002",
            name: "Premium Rose Arrangement",
            price: "$79.99",
            priceValue: 79.99,
            currency: "USD",
            imageUrl: "https://images.unsplash.com/photo-1455659817273-f96807779a8a?w=400",
            vendor: "FTD",
            buyUrl: "https://www.ftd.com",
            city: "New York"
        ),
        FlowerProduct(
            productId: "mock_003",
            name: "Red Roses in Vase",
            price: "2500 ₽",
            priceValue: 2500,
            currency: "RUB",
            imageUrl: "https://images.unsplash.com/photo-1490750967868-88aa4486c946?w=400",
            vendor: "Флорист.ру",
            buyUrl: "https://www.florist.ru",
            city: "Москва"
        ),
    ]
}

File: backend/schemas/flower_product.py
"""
Flower Product Schema - for "Find Flowers" feature

Defines product card data from flower shop search results.
"""

from typing import Optional, List
from pydantic import BaseModel, Field


class ShopCard(BaseModel):
    """Product card for flower shop search results."""

    product_id: str = Field(..., description="Unique product identifier")
    name: str = Field(..., description="Product name")
    price: str = Field(..., description="Formatted price (e.g., '$49.99', '2500 ₽')")
    price_value: Optional[float] = Field(None, description="Numeric price for sorting")
    currency: str = Field(default="USD", description="Currency code: USD, CAD, RUB")
    image_url: Optional[str] = Field(None, description="Product image URL")
    vendor: str = Field(..., description="Store/vendor name")
    buy_url: str = Field(..., description="Direct link to purchase")
    city: Optional[str] = Field(None, description="City where product is available")


# Alias for backward compatibility
FlowerProduct = ShopCard


class FlowerSearchRequest(BaseModel):
    """Request body for flower product search."""

    flower_name: str = Field(..., description="Flower name to search")
    city: str = Field(..., description="City for delivery/search")
    region: str = Field(default="US", description="Geographic region: US, CA, RU")
    max_results: int = Field(default=10, ge=1, le=20, description="Max products to return")
    skip_cache: bool = Field(default=False, description="Bypass cache and force fresh search")


class FlowerSearchResponse(BaseModel):
    """Response for flower product search."""

    success: bool = True
    query: str = Field(..., description="Search query used")
    provider: str = Field(default="unknown", description="Provider used: yandex, floristone, fallback")
    products: List[ShopCard] = Field(default_factory=list)
    error: Optional[str] = None
    cached_at: Optional[str] = Field(default=None, description="ISO timestamp when results were cached")
